diff -u -r mediatek/files-5.4/arch/arm64/boot/dts/mediatek/mt7981.dtsi b/files-5.4/arch/arm64/boot/dts/mediatek/mt7981.dtsi
--- mediatek/files-5.4/arch/arm64/boot/dts/mediatek/mt7981.dtsi	2024-09-22 12:31:56.698693462 +0800
+++ b/files-5.4/arch/arm64/boot/dts/mediatek/mt7981.dtsi	2024-09-19 15:19:55.000000000 +0800
@@ -381,7 +381,6 @@
                 #address-cells = <1>;
                 #size-cells = <1>;
                 compatible = "mediatek,mt7981-ethsys",
-                		"mediatek,mt7986-ethsys",
                              "syscon";
                 reg = <0 0x15000000 0 0x1000>;
                 #clock-cells = <1>;
@@ -682,10 +681,5 @@
 		clocks = <&infracfg_ao CK_INFRA_DBG_CK>;
 		clock-names = "ice_dbg";
 	};
-	
-	wifi: wifi@18000000 {
-		compatible = "mediatek,mt7981-wmac";
-		status = "disabled";
-	};
 };
 #include "mt7981-clkitg.dtsi"
diff -u -r mediatek/files-5.4/arch/arm64/boot/dts/mediatek/mt7986a-bananapi-bpi-r3mini-emmc.dts b/files-5.4/arch/arm64/boot/dts/mediatek/mt7986a-bananapi-bpi-r3mini-emmc.dts
--- mediatek/files-5.4/arch/arm64/boot/dts/mediatek/mt7986a-bananapi-bpi-r3mini-emmc.dts	2024-09-19 03:33:58.084268000 +0800
+++ b/files-5.4/arch/arm64/boot/dts/mediatek/mt7986a-bananapi-bpi-r3mini-emmc.dts	2024-09-23 01:21:02.873595000 +0800
@@ -201,8 +201,8 @@
 		};
 	};
 
-	gmac1: mac@1 {
-		label = "wan";
+	gmac1: mac@1 { 
+	
 		compatible = "mediatek,eth-mac";
 		reg = <1>;
 		phy-mode = "2500base-x";
Only in mediatek/files-5.4/arch: image
Only in mediatek/files-5.4/drivers/net/ethernet: 1.patch
diff -u -r mediatek/files-5.4/drivers/net/ethernet/mediatek/mtk_eth_soc.c b/files-5.4/drivers/net/ethernet/mediatek/mtk_eth_soc.c
--- mediatek/files-5.4/drivers/net/ethernet/mediatek/mtk_eth_soc.c	2024-09-18 02:01:12.484565000 +0800
+++ b/files-5.4/drivers/net/ethernet/mediatek/mtk_eth_soc.c	2024-09-23 01:41:55.151214000 +0800
@@ -2467,11 +2467,16 @@
 
 static void mtk_dma_free(struct mtk_eth *eth)
 {
-	int i;
+	int i, j, txqs;
+	txqs = MTK_QDMA_TX_NUM;
+
+	for (i = 0; i < MTK_MAC_COUNT; i++) {
+		if (!eth->netdev[i])
+			continue;
+		for (j = 0; j < txqs; j++)
+			netdev_tx_reset_queue(netdev_get_tx_queue(eth->netdev[i], j));
+	}
 
-	for (i = 0; i < MTK_MAC_COUNT; i++)
-		if (eth->netdev[i])
-			netdev_reset_queue(eth->netdev[i]);
 	if ( !eth->soc->has_sram && eth->scratch_ring) {
 		dma_free_coherent(eth->dev,
 				  MTK_DMA_SIZE * sizeof(struct mtk_tx_dma),
diff -u -r mediatek/files-5.4/drivers/net/ethernet/mediatek/mtk_eth_soc.h b/files-5.4/drivers/net/ethernet/mediatek/mtk_eth_soc.h
--- mediatek/files-5.4/drivers/net/ethernet/mediatek/mtk_eth_soc.h	2024-09-16 23:17:47.298331000 +0800
+++ b/files-5.4/drivers/net/ethernet/mediatek/mtk_eth_soc.h	2024-09-23 01:43:04.799378000 +0800
@@ -1165,12 +1165,12 @@
 
 #define MT7986_CAPS   (MTK_GMAC1_SGMII | MTK_GMAC2_SGMII | \
                        MTK_MUX_GMAC12_TO_GEPHY_SGMII | MTK_QDMA | MTK_NETSYS_RX_V2 | \
-                       MTK_NETSYS_V2)
+                       MTK_NETSYS_V2 | MTK_RSTCTRL_PPE1)
 
 #define MT7981_CAPS   (MTK_GMAC1_SGMII | MTK_GMAC2_SGMII | MTK_GMAC2_GEPHY | \
 			MTK_MUX_GMAC12_TO_GEPHY_SGMII | MTK_QDMA | \
 			MTK_MUX_U3_GMAC2_TO_QPHY | MTK_U3_COPHY_V2 | \
-			MTK_NETSYS_V2)
+			MTK_NETSYS_V2 | MTK_RSTCTRL_PPE1)
 
 /* struct mtk_eth_data -	This is the structure holding all differences
  *				among various plaforms
diff -u -r mediatek/files-5.4/drivers/net/ethernet/mediatek/mtk_hnat/hnat_nf_hook.c b/files-5.4/drivers/net/ethernet/mediatek/mtk_hnat/hnat_nf_hook.c
--- mediatek/files-5.4/drivers/net/ethernet/mediatek/mtk_hnat/hnat_nf_hook.c	2024-09-19 03:36:19.698275000 +0800
+++ b/files-5.4/drivers/net/ethernet/mediatek/mtk_hnat/hnat_nf_hook.c	2024-09-23 01:45:39.670194000 +0800
@@ -726,6 +726,12 @@
 	skb_hnat_alg(skb) = val;
 }
 
+static inline void hnat_set_tag(const struct nf_hook_state *state,
+				struct sk_buff *skb, int val)
+{
+	skb_hnat_magic_tag(skb) = val;
+}
+
 static inline void hnat_set_head_frags(const struct nf_hook_state *state,
 				       struct sk_buff *head_skb, int val,
 				       void (*fn)(const struct nf_hook_state *state,
@@ -991,8 +997,8 @@
 		goto drop;
 
 	if (!IS_WHNAT(state->in) && IS_EXT(state->in) && IS_SPACE_AVAILABLE_HEAD(skb)) {
-		skb_hnat_alg(skb) = 0;
-		skb_hnat_magic_tag(skb) = HNAT_MAGIC_TAG;
+		hnat_set_head_frags(state, skb, 0, hnat_set_alg);
+		hnat_set_head_frags(state, skb, HNAT_MAGIC_TAG, hnat_set_alg);
 	}
 
 	if (!is_magic_tag_valid(skb))
@@ -1065,8 +1071,8 @@
 		
 
 	if (!IS_WHNAT(state->in) && IS_EXT(state->in) && IS_SPACE_AVAILABLE_HEAD(skb)) {
-		skb_hnat_alg(skb) = 0;
-		skb_hnat_magic_tag(skb) = HNAT_MAGIC_TAG;
+		hnat_set_head_frags(state, skb, 0, hnat_set_alg);
+		hnat_set_head_frags(state, skb, HNAT_MAGIC_TAG, hnat_set_alg);
 	}
 
 	if (!is_magic_tag_valid(skb))
@@ -1127,8 +1133,8 @@
 		goto drop;
 	
 	if (!IS_WHNAT(state->in) && IS_EXT(state->in) && IS_SPACE_AVAILABLE_HEAD(skb)) {
-		skb_hnat_alg(skb) = 0;
-		skb_hnat_magic_tag(skb) = HNAT_MAGIC_TAG;
+		hnat_set_head_frags(state, skb, 0, hnat_set_alg);
+		hnat_set_head_frags(state, skb, HNAT_MAGIC_TAG, hnat_set_alg);
 	}
 
 	if (!is_magic_tag_valid(skb))
Only in mediatek/files-5.4/drivers/net/ethernet: mediatek_sdk
diff -u -r mediatek/files-5.4/net/nat/foe_hook/hook_base.c b/files-5.4/net/nat/foe_hook/hook_base.c
--- mediatek/files-5.4/net/nat/foe_hook/hook_base.c	2024-09-19 05:06:16.838066488 +0800
+++ b/files-5.4/net/nat/foe_hook/hook_base.c	2024-09-19 15:19:55.000000000 +0800
@@ -17,5 +17,3 @@
 
 int (*ra_sw_nat_hook_tx)(struct sk_buff *skb, int gmac_no) = NULL;
 EXPORT_SYMBOL(ra_sw_nat_hook_tx);
-
-
diff -u -r mediatek/Makefile b/Makefile
--- mediatek/Makefile	2024-07-29 00:58:26.655384000 +0800
+++ b/Makefile	2024-09-23 01:20:34.268917000 +0800
@@ -14,6 +14,6 @@
 include $(INCLUDE_DIR)/target.mk
 DEFAULT_PACKAGES += \
 	kmod-leds-gpio kmod-gpio-button-hotplug \
-	autocore-arm uboot-envtool luci-app-wrtbwmon
+	autocore-arm uboot-envtool luci-app-wrtbwmon mtk-smp
 
 $(eval $(call BuildTarget))
diff -u -r mediatek/modules.mk b/modules.mk
--- mediatek/modules.mk	2024-07-29 00:58:26.690451000 +0800
+++ b/modules.mk	2024-09-19 17:03:11.000000000 +0800
@@ -42,7 +42,7 @@
 define KernelPackage/mediatek_hnat
   SUBMENU:=Network Devices
   TITLE:=Mediatek HNAT module
-  DEPENDS:=@TARGET_mediatek +kmod-nf-conntrack
+  DEPENDS:=@TARGET_mediatek +kmod-nf-conntrack +wireless-tools +@KERNEL_WIRELESS_EXT
   AUTOLOAD:=$(call AutoLoad,20,mtkhnat)
   MODPARAMS.mtkhnat:=ppe_cnt=2
   KCONFIG:= \
diff -u -r mediatek/mt7981/config-5.4 b/mt7981/config-5.4
--- mediatek/mt7981/config-5.4	2024-07-29 00:58:26.690451000 +0800
+++ b/mt7981/config-5.4	2024-09-19 17:06:33.000000000 +0800
@@ -2,6 +2,7 @@
 CONFIG_AHCI_MTK=y
 # CONFIG_AIROHA_EN8801SC_PHY is not set
 # CONFIG_AIROHA_EN8801S_PHY is not set
+# CONFIG_AIR_EN8811H_PHY is not set
 CONFIG_ARCH_CLOCKSOURCE_DATA=y
 CONFIG_ARCH_DMA_ADDR_T_64BIT=y
 CONFIG_ARCH_KEEP_MEMBLOCK=y
@@ -50,6 +51,7 @@
 CONFIG_ARM_MEDIATEK_CPUFREQ=y
 CONFIG_ARM_PMU=y
 CONFIG_ARM_PSCI_FW=y
+# CONFIG_ASN1 is not set
 CONFIG_ATA=y
 CONFIG_AUDIT_ARCH_COMPAT_GENERIC=y
 CONFIG_BLK_DEV_DM=y
@@ -79,6 +81,7 @@
 CONFIG_BT_MTKUART=y
 CONFIG_BT_QCA=y
 CONFIG_CAVIUM_TX2_ERRATUM_219=y
+CONFIG_CC_HAVE_STACKPROTECTOR_SYSREG=y
 CONFIG_CLKDEV_LOOKUP=y
 CONFIG_CLKSRC_MMIO=y
 CONFIG_CLOCK_THERMAL=y
@@ -123,6 +126,7 @@
 CONFIG_CRYPTO_AUTHENC=y
 CONFIG_CRYPTO_CBC=y
 CONFIG_CRYPTO_CMAC=y
+CONFIG_CRYPTO_CRC32=y
 CONFIG_CRYPTO_CRYPTD=y
 CONFIG_CRYPTO_DEFLATE=y
 CONFIG_CRYPTO_DES=y
@@ -283,8 +287,8 @@
 # CONFIG_MEDIATEK_2P5GE_PHY is not set
 # CONFIG_MEDIATEK_GE_PHY is not set
 CONFIG_MEDIATEK_MT6577_AUXADC=y
-CONFIG_MEDIATEK_NETSYS_V2=y
 # CONFIG_MEDIATEK_NETSYS_RX_V2 is not set
+CONFIG_MEDIATEK_NETSYS_V2=y
 CONFIG_MEDIATEK_WATCHDOG=y
 CONFIG_MEDIA_SUPPORT=y
 CONFIG_MEMFD_CREATE=y
@@ -398,7 +402,6 @@
 CONFIG_PINCTRL_MT8516=y
 CONFIG_PINCTRL_MTK=y
 CONFIG_PINCTRL_MTK_MOORE=y
-CONFIG_PLUGIN_HOSTCC="g++"
 CONFIG_PM=y
 CONFIG_PM_CLK=y
 CONFIG_PM_GENERIC_DOMAINS=y
diff -u -r mediatek/mt7986/base-files/etc/board.d/02_network b/mt7986/base-files/etc/board.d/02_network
--- mediatek/mt7986/base-files/etc/board.d/02_network	2024-09-19 03:33:58.092268000 +0800
+++ b/mt7986/base-files/etc/board.d/02_network	2024-09-23 01:21:39.084984000 +0800
@@ -23,7 +23,7 @@
 		ucidef_set_interfaces_lan_wan "eth1 lan1 lan2 lan3" wan
 		;;
 	bananapi,bpi-r3mini*)
-		ucidef_set_interfaces_lan_wan "eth0" "eth1"
+		ucidef_set_interfaces_lan_wan "eth0" eth1
 		;;
 	netcore,n60 |\
 	ruijie,rg-x60-pro*)
diff -u -r mediatek/patches-5.4/999-2903-mtkhnat-enhance.patch b/patches-5.4/999-2903-mtkhnat-enhance.patch
--- mediatek/patches-5.4/999-2903-mtkhnat-enhance.patch	2024-09-14 18:55:28.789292000 +0800
+++ b/patches-5.4/999-2903-mtkhnat-enhance.patch	2024-09-23 08:47:00.525862000 +0800
@@ -1,5 +1,5 @@
 --- a/include/net/netfilter/nf_flow_table.h	2024-09-13 16:35:08.862364500 +0800
-+++ b/include/net/netfilter/nf_flow_table.h	2024-09-14 10:58:23.716107006 +0800
++++ b/include/net/netfilter/nf_flow_table.h	2024-09-14 10:28:23.716107006 +0800
 @@ -109,6 +109,7 @@
  	u16 vlan_id;
  	u16 pppoe_sid;
